pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,181 @@
+	}
+	if cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef == nil {
+		return fmt.Errorf("must specify machineGroupRef for control plane")
+	}
+
+	if cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef == nil {
+		return fmt.Errorf("must specify machineGroupRef for worker nodes")
+	}
+
+	controlPlaneMachineConfig := cloudStackClusterSpec.controlPlaneMachineConfig()
+	if controlPlaneMachineConfig == nil {
+		return fmt.Errorf("cannot find CloudStackMachineConfig %v for control plane", cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name)
+	}
+
+	if cloudStackClusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if cloudStackClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef == nil {
+			return fmt.Errorf("must specify machineGroupRef for etcd machines")
+		}
+		etcdMachineConfig = cloudStackClusterSpec.etcdMachineConfig()
+		if etcdMachineConfig == nil {
+			return fmt.Errorf("cannot find CloudStackMachineConfig %v for etcd machines", cloudStackClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name)
+		}
+		if etcdMachineConfig.Spec.Template != controlPlaneMachineConfig.Spec.Template {
+			return fmt.Errorf("control plane and etcd machines must have the same template specified")
+		}
+	}
+
+	if cloudStackClusterSpec.datacenterConfig.Namespace != cloudStackClusterSpec.Namespace {
+		return fmt.Errorf(
+			"CloudStackDatacenterConfig and Cluster objects must have the same namespace: CloudstackDatacenterConfig namespace=%s; Cluster namespace=%s",
+			cloudStackClusterSpec.datacenterConfig.Namespace,
+			cloudStackClusterSpec.Namespace,
+		)
+	}
+
+	workerNodeGroupMachineConfig, ok := v.machineConfigs[cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name]
+	if !ok {
+		return fmt.Errorf("cannot find CloudStackMachineConfig %v for worker nodes", cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name)
+	}
+	if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+		if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+			return fmt.Errorf("control plane and worker nodes must have the same template specified")
+		}
+	}
+
+	hostWithPort, err := v.validateControlPlaneHostAndApplyDefaultPort(cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host)
+	if err != nil {
+		return fmt.Errorf("failed to validate controlPlaneConfiguration.Endpoint.Host: %v", err)
+	}
+	cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host = hostWithPort
+
+	for _, machineConfig := range v.machineConfigs {
+		if machineConfig.Namespace != cloudStackClusterSpec.Namespace {
+			return fmt.Errorf(
+				"CloudStackMachineConfig and Cluster objects must have the same namespace: CloudStackMachineConfig namespace=%s; Cluster namespace=%s",
```

Comment:
It might be easier for the user to correct this mistake if we include the name of the invalid `CloudStackMachineConfig`